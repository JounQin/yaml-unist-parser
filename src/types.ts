export interface Point {
  /** 1-based */
  line: number;
  /** 1-based */
  column: number;
  /** 0-based */
  offset: number;
}

export interface Position {
  start: Point;
  end: Point;
}

export interface Node {
  type: string;
  position: Position;
}

export interface Parent extends Node {
  children: Node[];
}

// -----------------------------------------------------------------------------

export interface Content {
  anchor?: string;
  tag?: { verbatim: string } | { handle: string; suffix: string };
}

export interface CommentAttachable {
  /** comments in front of the node */
  leadingComments: Comment[];
  /** comments between the node and its tag/anchor */
  middleComments: Comment[];
  /** comments on the same line of the node */
  trailingComments: Comment[];
}

export interface YAMLSyntaxError extends SyntaxError {
  source: string;
  position: Position;
}

export type YamlUnistNode =
  | Comment
  | Root
  | Document
  | DocumentHead
  | DocumentBody
  | Directive
  | Alias
  | BlockLiteral
  | BlockFolded
  | Plain
  | QuoteSingle
  | QuoteDouble
  | Mapping
  | MappingItem
  | MappingKey
  | MappingValue
  | Sequence
  | SequenceItem
  | FlowMapping
  | FlowSequence
  | Null;

export type ContentNode =
  | Alias
  | BlockLiteral
  | BlockFolded
  | Plain
  | QuoteSingle
  | QuoteDouble
  | Mapping
  | Sequence
  | FlowMapping
  | FlowSequence
  | Null;

// -----------------------------------------------------------------------------

export interface Null extends Node {
  type: "null";
  /** always -1 */
  position: Position;
}

export interface Comment extends Node {
  type: "comment";
  value: string;
  /** @internal non-enumerable */
  parent?: Extract<YamlUnistNode, CommentAttachable>;
}

export interface Root extends Parent {
  type: "root";
  children: Document[];
  comments: Comment[];
}

export interface Document extends Parent {
  type: "document";
  children: [DocumentHead, DocumentBody];
}

export interface DocumentHead extends Parent {
  type: "documentHead";
  children: Array<Comment | Directive>;
}

export interface DocumentBody extends Parent {
  type: "documentBody";
  children: Array<Comment | ContentNode>;
}

export interface Directive extends Node, CommentAttachable {
  type: "directive";
  name: string;
  parameters: string[];
}

export interface Alias extends Node, CommentAttachable {
  type: "alias";
  value: string;
}

export interface BlockValue extends Node, Content, CommentAttachable {
  chomping: "clip" | "keep" | "strip";
  indent?: number;
  value: string;
}

export interface BlockLiteral extends BlockValue {
  type: "blockLiteral";
}

export interface BlockFolded extends BlockValue {
  type: "blockFolded";
}

export interface Plain extends Node, Content, CommentAttachable {
  type: "plain";
  value: string;
}

export interface QuoteValue extends Node, Content, CommentAttachable {
  value: string;
}

export interface QuoteSingle extends QuoteValue {
  type: "quoteSingle";
}

export interface QuoteDouble extends QuoteValue {
  type: "quoteDouble";
}

export interface Mapping extends Parent, Content, CommentAttachable {
  type: "mapping";
  children: MappingItem[];
}

export interface MappingItem extends Parent, CommentAttachable {
  type: "mappingItem";
  /** key-value pair */
  children: [MappingKey, MappingValue];
}

export interface MappingKey extends Parent, CommentAttachable {
  type: "mappingKey";
  children: [ContentNode];
}

export interface MappingValue extends Parent, CommentAttachable {
  type: "mappingValue";
  children: [ContentNode];
}

export interface Sequence extends Parent, Content, CommentAttachable {
  type: "sequence";
  children: SequenceItem[];
}

export interface SequenceItem extends Parent, CommentAttachable {
  type: "sequenceItem";
  children: [ContentNode];
}

export interface FlowCollection extends Parent, Content, CommentAttachable {
  children: Array<MappingItem | SequenceItem>;
}

export interface FlowMapping extends FlowCollection {
  type: "flowMapping";
  children: MappingItem[];
}

export interface FlowSequence extends FlowCollection {
  type: "flowSequence";
  children: Array<MappingItem | SequenceItem>;
}
