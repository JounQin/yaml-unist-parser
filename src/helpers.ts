import { parse } from "./parse";
import {
  CommentAttachable,
  Node,
  Position,
  Root,
  YamlUnistNode,
} from "./types";

const RAW = Symbol("raw");

expect.addSnapshotSerializer({
  test: value => typeof value[RAW] === "string",
  print: value => value[RAW].replace(/ +$/gm, ""),
});

export type TestCase = TestCaseSingle | TestCaseMulti;
export type TestCaseSelector = (root: Root) => YamlUnistNode | Node;

export type TestCaseSingle = [string, TestCaseSelector];
export type TestCaseMulti = [string, TestCaseSelector[]];

export function getFirstContent<T extends YamlUnistNode>(): (root: Root) => T;
export function getFirstContent<T extends YamlUnistNode>(root: Root): T;
export function getFirstContent<T extends YamlUnistNode>(root?: Root) {
  return !root
    ? (_root: Root) => _root.children[0].children[1].children[0] as T
    : (root.children[0].children[1].children[0] as T);
}

export function testCases(cases: TestCase[]) {
  cases.forEach(testCase => {
    const [text, selector] = testCase;
    const selectNodes = ([] as TestCaseSelector[]).concat(selector);
    selectNodes.forEach(selectNode => {
      test(`${JSON.stringify(text)}`, () => {
        expect({ [RAW]: snapshotNode(text, selectNode) }).toMatchSnapshot();
      });
    });
  });
}

function snapshotNode(
  text: string,
  selectNode: (root: Root) => YamlUnistNode | Node,
) {
  const root = parse(text);
  const node = selectNode(root);
  return (
    `${node.type} (` +
    `${node.position.start.line}:${node.position.start.column} ~ ` +
    `${node.position.end.line}:${node.position.end.column}` +
    ")\n" +
    codeFrameColumns(text, node.position) +
    "\n" +
    stringifyNode(node)
  );
}

function stringifyNode(node: YamlUnistNode | Node): string {
  const attributes = Object.keys(node)
    .filter(key => {
      switch (key) {
        case "type":
        case "position":
        case "children":
        case "leadingComments":
        case "middleComments":
        case "trailingComments":
          return false;
        default:
          return true;
      }
    })
    .map(key => `${key}=${JSON.stringify(node[key as keyof typeof node])}`)
    .sort()
    .map((attribute, index) => (index === 0 ? " " + attribute : attribute))
    .join(" ");
  const comments =
    "leadingComments" in node
      ? (["leadingComments", "middleComments", "trailingComments"] as Array<
          keyof CommentAttachable
        >)
          .map(key =>
            node[key].map(
              comment =>
                `<${key.slice(0, -1)} value=${JSON.stringify(comment.value)}>`,
            ),
          )
          .reduce((a, b) => a.concat(b), [])
      : [];
  return "children" in node || comments.length !== 0
    ? `<${node.type}${attributes}>\n${indent(
        comments
          .concat(
            "children" in node
              ? (node.children as YamlUnistNode[]).map(stringifyNode)
              : [],
          )
          .join("\n"),
      )}\n</${node.type}>`
    : `<${node.type}${attributes} />`;
}

function indent(text: string) {
  return text
    .split("\n")
    .map(line => "  " + line)
    .join("\n");
}

function codeFrameColumns(text: string, position: Position) {
  const lines = text.split("\n").map(line => `${line}¶`);
  const markerLines = lines.map((line, index) => {
    if (index < position.start.line - 1 || index > position.end.line - 1) {
      return "";
    }
    if (index === position.start.line - 1) {
      return (
        " ".repeat(position.start.column - 1) +
        (index === position.end.line - 1 &&
        position.start.column === position.end.column
          ? "~"
          : "^".repeat(
              (index === position.end.line - 1
                ? position.end.column - 1
                : line.length) -
                (position.start.column - 1),
            ))
      );
    } else if (index === position.end.line - 1) {
      return position.end.column === 1 && line === "¶"
        ? "^"
        : "^".repeat(position.end.column - 1);
    } else {
      return "^".repeat(line.length);
    }
  });

  const gutterWidth = Math.floor(Math.log10(lines.length)) + 1;

  return lines
    .reduce(
      (reduced, line, index) => {
        const gutter = leftpad(`${index + 1}`, gutterWidth);
        return reduced.concat(
          `${gutter} | ${line.replace(/ /g, "·")}`,
          markerLines[index]
            ? `${" ".repeat(gutterWidth)} | ${markerLines[index]}`
            : [],
        );
      },
      [] as string[],
    )
    .join("\n");
}

function leftpad(text: string, width: number) {
  return " ".repeat(Math.max(0, width - text.length)) + text;
}
