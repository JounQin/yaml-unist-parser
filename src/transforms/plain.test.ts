import { getFirstContent, testCases } from "../helpers";
import { Plain } from "../types";

testCases([["   123   ", getFirstContent()]]);
testCases([
  [
    " !!str # comment 1 \n &anchor # comment 2 \n  123   ",
    [
      getFirstContent(),
      root => getFirstContent<Plain>(root).middleComments[0],
      root => getFirstContent<Plain>(root).middleComments[1],
    ],
  ],
]);
