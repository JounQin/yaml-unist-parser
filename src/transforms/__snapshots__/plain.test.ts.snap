// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"   123   " 1`] = `
plain (1:4 ~ 1:7)
1 | ···123···¶
  |    ^^^
<plain value="123" />
`;

exports[`" !!str # comment 1 \\n &anchor # comment 2 \\n  123   " 1`] = `
plain (1:2 ~ 3:6)
1 | ·!!str·#·comment·1·¶
  |  ^^^^^^^^^^^^^^^^^^^
2 | ·&anchor·#·comment·2·¶
  | ^^^^^^^^^^^^^^^^^^^^^^
3 | ··123···¶
  | ^^^^^
<plain anchor="anchor" tag={"handle":"!!","suffix":"str"} value="123">
  <middleComment value=" comment 1 ">
  <middleComment value=" comment 2 ">
</plain>
`;

exports[`" !!str # comment 1 \\n &anchor # comment 2 \\n  123   " 2`] = `
comment (1:8 ~ 1:20)
1 | ·!!str·#·comment·1·¶
  |        ^^^^^^^^^^^^
2 | ·&anchor·#·comment·2·¶
3 | ··123···¶
<comment value=" comment 1 " />
`;

exports[`" !!str # comment 1 \\n &anchor # comment 2 \\n  123   " 3`] = `
comment (2:10 ~ 2:22)
1 | ·!!str·#·comment·1·¶
2 | ·&anchor·#·comment·2·¶
  |          ^^^^^^^^^^^^
3 | ··123···¶
<comment value=" comment 2 " />
`;
